src/sieve.py

```diff
@@ -14,7 +17,19 @@ def sieve(n: int) -> list[int]:
     assert n > 0
     candidates = list(range(2, n + 1))
     primes = []
-
+    # 2 is a prime number
+    primes.append(2)
+    #iterate over 3 to n+1
+    for candidate in list(range(3, n + 1)):
+        #calculate the residuals --> if any number is divisible (residual == 0), then it is not prime
+        residuals = [candidate % p for p in primes]
+         
+        if 0 not in residuals: #prime number
+            primes.append(candidate)
+        else: #composite number
+            candidates.remove(candidate)
```

Comment:
I really like this solution, as it does the filtering in a different order than what I am used to. Normally, the sieve will eliminate candidates one prime at a time, but this one instead checks each candidate against all the previous  primes, and that is an excellent solution. (The usual solution can be optimised a little compared to this, but we don't care about efficiency until next week).

I would clean it up a little myself. You don't need the `candidates` list, when you solve the problem this way. You only need the list of `primes`. So you could rewrite the function as something like this:

```python
def sieve(n: int) -> list[int]:
    """Find primes."""
    primes = [2]
    for candidate in range(3, n + 1):
        residuals = [candidate % p for p in primes]
        if 0 not in residuals:
            primes.append(candidate)
    return primes
```

The `0 in residuals` requires that we first compute all the residuals and then check if there is a zero in it, but we know the answer we are looking for if we see the first one, so we could also use the function `any` that stops once it knows the answer. That would look like this:

```python
def sieve(n: int) -> list[int]:
    """Find primes."""
    primes = [2]
    for candidate in range(3, n + 1):
        if not any(candidate % p == 0 for p in primes):
            primes.append(candidate)
    return primes
```

If you don't like the `if not any(...)` that can be hard to read with the negations, there is a corresponding `all` function you can use instead:

```python
def sieve(n: int) -> list[int]:
    """Find primes."""
    primes = [2]
    for candidate in range(3, n + 1):
        if all(candidate % p != 0 for p in primes):
            primes.append(candidate)
    return primes
```

You could also write a loop for checking residuals and exploit that the `else` part of a loop in Python is only run if you make it all the way through the loop without `break`'ing.

```python
def sieve(n: int) -> list[int]:
    """Find primes."""
    primes = [2]
    for candidate in range(3, n + 1):
        for p in primes:
            if candidate % p == 0:
                break
        else:
            # This only runs if we didn't break
            primes.append(candidate)
    return primes
```

Excellent solution!